navigation_and_routing/lib/src/routing/delegate.dart

```diff
@@ -0,0 +1,144 @@
+// Copyright 2021, the Flutter project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:bookstore/src/auth/auth.dart';
+import 'package:bookstore/src/screens/sign_in.dart';
+import 'package:flutter/foundation.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter/widgets.dart';
+import 'package:collection/collection.dart';
+
+import '../data/author.dart';
+import '../data/book.dart';
+import '../data/library_scope.dart';
+import '../screens/author_details.dart';
+import '../screens/book_details.dart';
+import '../screens/scaffold.dart';
+import '../util/fade_transition_page.dart';
+import 'guard.dart';
+import 'parsed_route.dart';
+import 'route_state.dart';
+
+/// Displays a [BookstoreScaffold] with either a [BookDetailsScreen] or
+/// [AuthorDetailsScreen] stacked on top.
+class BookstoreRouterDelegate extends RouterDelegate<ParsedRoute>
+    with ChangeNotifier, PopNavigatorRouterDelegateMixin<ParsedRoute> {
+  final RouteState routeState;
+  final BookstoreAuth auth;
+  final RouteGuard<ParsedRoute> guard;
+  final scaffoldKey = const ValueKey<String>('App scaffold');
+  final bookDetailsKey = const ValueKey<String>('Book details screen');
+  final authorDetailsKey = const ValueKey<String>('Author details screen');
+
+  @override
+  final GlobalKey<NavigatorState> navigatorKey = GlobalKey<NavigatorState>();
+
+  BookstoreRouterDelegate(this.routeState, this.auth, this.guard) {
+    routeState.addListener(notifyListeners);
+    auth.addListener(notifyListeners);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final pathTemplate = routeState.route.pathTemplate;
+    final library = LibraryScope.of(context);
+
+    Book? book;
+    if (pathTemplate == '/books/:bookId') {
```

Comment:
I wonder whether we should use ParsedRoute subclass to differentiate the route.

Reply:
directly comparing the pattern feel weird.

Reply:
My thinking was that `TemplateRouteParser` is a general-purpose class (not app-specific), so it returns a `ParsedRoute`, which is also general-purpose.

Reply:
After thinking about this some more, I do think it makes sense to look at the path template here. In this case, the same widget gets built for every path matching that pattern. Comparing the ParsedRoute would be for cases where you want to build a different widget depending on the exact path (`/book/1` and `/book/2` would build two different widgets)

Reply:
Maybe define the string as a const variable somewhere?